doc/userguide/rules/datasets.rst

```diff
@@ -0,0 +1,152 @@
+Datasets
+========
+
+Using the ``dataset`` and ``datarep`` keyword it is possible to match on
+large amounts of data against any sticky buffer.
+
+For example, to match against a dns black list called ``dns-bl``::
+
+    dns.query; dataset:isset,dns-bl;
+
+These keywords are aware of transforms. So to look up a DNS query again
+a MD5 black list::
+
+    dns.query; to_md5; dataset:isset,dns-bl;
+
+Global config (optional)
+------------------------
+
+Datasets can optionally be defined in the main config. Sets can also be
+declared from the rule syntax.
+
+Example of sets for tracking unique values::
+
+    datasets:
+     - ua-seen:
+       type: string
+       state: ua-seen.lst
+     - dns-sha256-seen:
+       type: sha256
+       state: dns-sha256-seen.lst
+
+Rules to go with the above::
+
+    http.user_agent; dataset:set,ua-seen;
+    dns.query; to_sha256; dataset:set,dns-sha256-seen;
+
+
+Rule keywords
+-------------
+
+dataset
+~~~~~~~
+
+Datasets are binary: something is in the set or it's not.
+
+Syntax::
+
+    dataset:<cmd>,<name>,<options>;
+
+    dataset:<set|isset|isnotset>,<name> \
+        [, type <string|md5|sha256>, save <file name>, load <file name>, state <file name>];
+
+type <type>
+  the data type: string, md5, sha256
+save <file name>
+  file name for saving the in memory data when Suricata exits
+load <file name>
+  file name for load the data when Suricata starts up
+state
+  sets both 'save' and 'load' to the same value
```

Comment:
When would you want to use a different combination of the save, load and state?

Reply:
Is an initial `load` file something we could see a rule distributor releasing?

Reply:
You can either use load, load and save, or state. Keyword parsing will block the rest.

Just a load would indeed be what I expect to be used by a rule vendor.

State would be used in case you use either rules to fill a set (dataset + set operator) or if you manage it over unix socket.

Save is something you can use to just extract data. E.g. get all unique UAs.

Reply:
I wonder if the default directories should differ for a set vs, state. Think of the case where a ruleset tarball contains a rule, and an associated dataset, I think the best place to put that dataset is in the rule directory itself, or in a sub-directory of the rule directory named `datasets`.  In fact, that might be a good convention for rule publishers to follow as well.

The rule would then contain `dataset: load datasets/myset.lst`.  The rule parser would then look for this relative to the directory the rule is being loaded from.

Its a little less clear what to do with save, and state, and its likely those won't be used by rule publishers.